tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ConfidentialClientIntegrationTests.cs

```diff
@@ -647,6 +712,71 @@ private async Task RunOnBehalfOfTestAsync(LabResponse labResponse)
+            Assert.AreEqual(atHash, userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+            Assert.AreEqual(TokenSource.Cache, authResult.AuthenticationResultMetadata.TokenSource);
+
+            //Expire access tokens
+            TokenCacheHelper.ExpireAccessTokens(confidentialApp.UserTokenCacheInternal);
+
+            //Run OBO again. Should do token refresh since the AT is expired
+            authResult = await confidentialApp.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
+            Assert.IsNotNull(authResult.AccessToken);
+            Assert.IsTrue(!userCacheRecorder.LastAfterAccessNotificationArgs.IsApplicationCache);
+            Assert.IsTrue(userCacheRecorder.LastAfterAccessNotificationArgs.HasTokens);
+            Assert.AreEqual(atHash, userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+            Assert.AreEqual(TokenSource.IdentityProvider, authResult.AuthenticationResultMetadata.TokenSource);
+
+            //creating second app with no refresh tokens
+            var atItems = await confidentialApp.UserTokenCacheInternal.GetAllAccessTokensAsync(true).ConfigureAwait(false);
+            var confidentialApp2 = ConfidentialClientApplicationBuilder
+                .Create(confidentialClientID)
+                .WithAuthority(new Uri(oboHost + authResult.TenantId), true)
+                .WithClientSecret(secret)
+                .WithTestLogging()
+                .BuildConcrete();
+
+            TokenCacheHelper.ExpireAndSaveAccessToken(confidentialApp2.UserTokenCacheInternal, atItems.FirstOrDefault());
+
+            //Should perform OBO flow since the access token is expired and the refresh token does not exist
+            authResult = await confidentialApp2.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
+            Assert.IsNotNull(authResult.AccessToken);
+            Assert.IsTrue(!userCacheRecorder.LastAfterAccessNotificationArgs.IsApplicationCache);
+            Assert.IsTrue(userCacheRecorder.LastAfterAccessNotificationArgs.HasTokens);
+            Assert.AreEqual(atHash, userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+            Assert.AreEqual(TokenSource.IdentityProvider, authResult.AuthenticationResultMetadata.TokenSource);
+
+            TokenCacheHelper.ExpireAccessTokens(confidentialApp2.UserTokenCacheInternal);
+            TokenCacheHelper.UpdateUserAssertions(confidentialApp2);
+
+            //Should perform OBO flow since the access token and the refresh token contains the wrong user assertion hash
+            authResult = await confidentialApp2.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
```

Comment:
if the hash is wrong it shouldn't work, this would mean i could use any userAssertion

Reply:
If the hash is wrong I think it should ignore the tokens in the cache and try to fetch new tokens from the endpoint with the new userAssertion. If the userAssertion is determined to not be valid by AAD then no tokens will be acquired.

Reply:
if the cache is partitioned, how can we have the wrong hash? I think either 1) the comment is not right or 2) the test is not right.